cpu/stm32f4/vectors.c

```diff
@@ -22,6 +22,19 @@
 
 #include "vectors_cortexm.h"
 
+#ifdef MODULE_PUF_SRAM
+#include "puf_sram.h"
+
+/* SRAM memory marker defined in the linker script */
+extern uint32_t _srelocate;
+
+void pre_startup(void) {
```

Comment:
please don't use ```pre_startup()```. Some platforms need that for pre-initialization (kinetis, nrf51).
For puf_sram, IMO it is valid to add an ```#ifdef MODULE_PUF_SRAM``` where needed, e.g., for cortexm, in cpu/cortexm/vectors_cortexm.c just before or after the call to ```pre_startup()```.

Reply:
OK.

Then there is the issue of those twelve duplicated lines... Can you factor them out?

Reply:
I could make one function call from:

```
if (!puf_sram_softreset()) {
    puf_sram_generate((uint8_t *)&_srelocate, SEED_RAM_LEN);
}
```
but the rest won't change -> I don't see the benefit. Furthermore, in the current state the RAM start address and length are similar for the three MCUs , but this might vary for different boards.

Reply:
How much do you expect the implementations differ?

If you move all of the current (three times duplicated) ```pre_startup()``` into ```sys/puf_sram/puf_sram.c``` and rename it to ```puf_sram_init()```, the duplication is removed. You can even move ```extern uint32_t _srelocate;``` into the function. Call that function from the cortexm reset_vector_default(). That would probably work on *all* cortexm boards. Just to be sure, it makes sense to add the feature to FEATURES_PROVIDED one-by-one, on a case-by-case basis.

Reply:
>If you move all of the current (three times duplicated) pre_startup() into sys/puf_sram/puf_sram.c and rename it to puf_sram_init(), the duplication is removed. 

@kaspar030 can we postpone a proper refactoring to a follow-up? I am not very convinced of putting all above lines into a generic `puf_sram_init()` function and expect it to work for all cortex platforms. Besides that, there's no experience with other platforms (avr, mips, ...) yet and there might be other parameters to consider for `puf_sram_init()`. I propose that we first think about the API and which abstractions we need in order to make this module platform-independent and *then* do the proper refactoring. Would that be okay? We should also mark this module in the meantime as `@experimental` (see #8948)

Reply:
> I am not very convinced of putting all above lines into a generic puf_sram_init() function and expect it to work for all cortex platforms. 

It doesn't have to work on more than the three mentioned platforms where it is made available through ```FEATURES_PROVIDED```.

> Besides that, there's no experience with other platforms (avr, mips, ...) yet and there might be other parameters to consider for puf_sram_init(). 

Irrelevant for now. If ```puf_sram_init()```  needs to be adapted for other architectures, we can introduce changes or multiple versions *then*.

> I propose that we first think about the API and which abstractions we need in order to make this module platform-independent and then do the proper refactoring. Would that be okay? 

IMO it is OK to merge half-experimental code into the main tree, given the promise that it gets fixed quickly. But the experimental part then has to be contained as much as possible.

Here that is not the case, 3*12 duplicate lines are introduced in cpu/, for *no reason*. Just copying the given ```pre_startup()``` into one file would lead to identical code without the duplication. That *cannot* be too much to ask to fulfill point 2.1. of the reviewing guidelines.